### PR TITLE
Accelerate-Christina S.

### PR DESCRIPTION
**Pull request by Peacegypsy**

### CODE REVIEW

File: viewing_party/party.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+def create_movie(title, genre, rating):
+    m = dict();
```
Comment (anselrognlie) [inline viewing_party/party.py:2]: Variable names like `m` are ok for counter variables, but try to provide more meaningful variable names for other use cases.  In this case you might consider naming it `movie`.
Python allows but does not encourage semicolons.
```suggestion
    movie = dict()
```

File: viewing_party/party.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    m['title'] = title
+    m['genre'] = genre
+    m['rating'] = rating
+    if m['title'] and m['genre'] and m['rating']:
+        return m
```
Comment (anselrognlie) [inline viewing_party/party.py:7]: This does result in the specified behavior (returning the new movie as long as none of the fields are empty, and `None` otherwise), but it's surprising to have a function that `return`s along one code path, and does not along another. We could check for valid inputs before doing any "real" work (explicitly returning `None`) and returning the populated movie otherwise. Consider:

```python
    # start of function
    if not title or not genre or not rating:
        return None

    # code to populate movie...

    return movie
```

File: viewing_party/party.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    
+def add_to_watched(user_data, movie):
```
Comment (anselrognlie) [inline viewing_party/party.py:9]: 👍

File: viewing_party/party.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    user_data['watched'].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
```
Comment (anselrognlie) [inline viewing_party/party.py:13]: 👍

File: viewing_party/party.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    user_data['watchlist'].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data['watchlist']:
+        if title in movie['title']:
```
Comment (anselrognlie) [inline viewing_party/party.py:19]: Be careful using `in` here. It performs a substring search rather than an equality comparison.
```suggestion
        if title == movie['title']:
```

File: viewing_party/party.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            user_data['watchlist'].remove(movie)
```
Comment (anselrognlie) [inline viewing_party/party.py:20]: We haven't discussed this in the curriculum, but it's a _little_ dangerous to remove things from, or add things to, a list while we're iterating over it. Here it's ok, since we immediately stop the iteration after having removed the movie we were looking for. But if we were attempting multiple removals, there can be strange side-effects.

File: viewing_party/party.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            add_to_watched(user_data, movie)
```
Comment (anselrognlie) [inline viewing_party/party.py:21]: Nice function re-use!

File: viewing_party/party.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            return user_data
```
Comment (anselrognlie) [inline viewing_party/party.py:22]: Rather than duplicating the `return user_data`, we could `break` out of the loop here. That would properly halt the iteration, and then run the following `return user_data`.

File: viewing_party/party.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    return user_data
+    
+def get_watched_avg_rating(user_data):
```
Comment (anselrognlie) [inline viewing_party/party.py:25]: 👍 

Formatting-wise, consider using blank lines to separate chunks of related code, especially after conditions and loops.

File: viewing_party/party.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    for movie in user_data['watched']:
+        movie_rating_total += movie['rating']
+    return movie_rating_total / len(user_data['watched'])    
+
+def get_most_watched_genre(user_data):
+    movie_genre_dictionary = {}
+    if len(user_data['watched']) < 1:
+        return None
+    for movie in user_data['watched']:
+        if movie['genre'] not in movie_genre_dictionary.keys():
```
Comment (anselrognlie) [inline viewing_party/party.py:38]: `in` for a dictionary already looks in the dictionary keys. We don't need to call `.keys()`
```suggestion
        if movie['genre'] not in movie_genre_dictionary:
```

File: viewing_party/party.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            movie_genre_dictionary[movie['genre']] = 1 
+
```
Comment (anselrognlie) [inline viewing_party/party.py:40]: Here, I'd avoid the blank line to help keep both parts of the `if` condition together.

File: viewing_party/party.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        else:
+            movie_genre_dictionary[movie['genre']] += 1
+
+    most_watched = sorted(movie_genre_dictionary, key=movie_genre_dictionary.get)
```
Comment (anselrognlie) [inline viewing_party/party.py:44]: Another standard method we could use is `max`.

Also, how might we find this result without a built-in function?

File: viewing_party/party.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    
+def get_watched_avg_rating(user_data):
+    if len(user_data['watched']) < 1:
+        return 0.0
+    movie_rating_total = 0
+    for movie in user_data['watched']:
+        movie_rating_total += movie['rating']
+    return movie_rating_total / len(user_data['watched'])    
+
+def get_most_watched_genre(user_data):
```
Comment (anselrognlie) [inline viewing_party/party.py:33]: 👍

File: viewing_party/party.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        if movie['genre'] not in movie_genre_dictionary.keys():
+            movie_genre_dictionary[movie['genre']] = 1 
+
+        else:
+            movie_genre_dictionary[movie['genre']] += 1
+
+    most_watched = sorted(movie_genre_dictionary, key=movie_genre_dictionary.get)
+    return most_watched[-1]  
+   
+def get_unique_watched(user_data):
```
Comment (anselrognlie) [inline viewing_party/party.py:47]: 👍

File: viewing_party/party.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    unique_list = []
+    for movie in user_data['watched']:
+        user_movie_list.append(movie['title'])
+
+    for friend in user_data['friends']:
+        for title in friend['watched']:
+            friends_movie_list.append(title['title'])
+
+    for m in user_movie_list:
+        if m not in friends_movie_list:
```
Comment (anselrognlie) [inline viewing_party/party.py:59]: The `in` (`not in`) check for a list is a O(n) operation. Could we use a different data structure that would let us look up the friend movie data more efficiently? Are there other places we could do this?

File: viewing_party/party.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        else:
+            movie_genre_dictionary[movie['genre']] += 1
+
+    most_watched = sorted(movie_genre_dictionary, key=movie_genre_dictionary.get)
+    return most_watched[-1]  
+   
+def get_unique_watched(user_data):
+    user_movie_list = []
+    friends_movie_list = []
+    unique_list = []
```
Comment (anselrognlie) [inline viewing_party/party.py:50]: Consider initializing variables closer to their first use. This can help us avoid accidentally using the wrong variable if there are several that are similarly named.

File: viewing_party/party.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+def get_friends_unique_watched(user_data):
+    user_movie_list = []
+    friends_movie_list = []
+    unique_list = []
+    for movie in user_data['watched']:
+        user_movie_list.append(movie['title'])
+
+    for friend in user_data['friends']:
+        for title in friend['watched']:
+            friends_movie_list.append(title['title'])
```
Comment (anselrognlie) [inline viewing_party/party.py:74]: These are identical to code in the previous method. How could we avoid this repetition?

File: viewing_party/party.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+            
+    for m in friends_movie_list:
+        if m not in user_movie_list:
+            if {'title': m } not in unique_list:
+                unique_list.append({'title': m })
+            else:
+                continue
```
Comment (anselrognlie) [inline viewing_party/party.py:81]: `continue` jumps back up to the start of the loop, but we're already at the end of the loop here. This `else` isn't necessary.

File: viewing_party/party.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            
+    return unique_list
+
+def get_available_recs(user_data):
+    recommended_movies = []
+    user_subscriptions = user_data['subscriptions']
+    user_watched = user_data['watched']
```
Comment (anselrognlie) [inline viewing_party/party.py:88]: `user_watched` isn't used in this method, but it probably should. There weren't any tests for cases where a potential recommendation was already in the user's watched list, but that was specified in the function description.

File: viewing_party/party.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        if item['host'] in user_subscriptions:
+            if item not in recommended_movies:
+                recommended_movies.append(item)
+
+    return recommended_movies
+
+def get_new_rec_by_genre(user_data):
+    recs_by_genre_list = []
+    potential_list = []      
+    genre_to_find = get_most_watched_genre(user_data)
```
Comment (anselrognlie) [inline viewing_party/party.py:105]: Nice function reuse!

File: viewing_party/party.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+    for item in (user_data['friends']):
+        for i in item['watched']:
+            potential_list.append(i)
+    for movie in potential_list:
```
Comment (anselrognlie) [inline viewing_party/party.py:110]: This wasn't listed as a requirement, but what would happen if two friends had the same film in the user's preferred genre? Do you have a method that could help avoid that situation?

File: viewing_party/party.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+            recs_by_genre_list.append(movie)
+    return recs_by_genre_list
+
+def get_rec_from_favorites(user_data):
+    user_fav_movies = []
+    recommended_movies = []
+    friends_movie_list = []
+    # potential_movies = get_friends_unique_watched(user_data)
+    for movie in user_data['favorites']:
+        user_fav_movies.append(movie)
```
Comment (anselrognlie) [inline viewing_party/party.py:121]: This looks to be copying the user's list of favorite movies, but couldn't we use `user_data['favorites']` directly?

**Review by anselrognlie — COMMENTED** [review]
Nice job! Your code is laid out well, and all the tests pass.

There are a couple of ares where tests didn't cover requirements or possible edge cases that you could think about.

One takeaway I'd like to emphasize is that `in` on lists is a O(n) operation, so if we're trying to generate a unique list, as the list gets longer, our `in` check gets slower and slower. Think about other data structures we could use to track whether we've seen an item before.
